### DISCUSSION TITLE
BC issue

### DISCUSSION THREAD

**Discussion by jin0123456** [idaholab/moose#25959] [Q&A General]
Hello,
![image](https://github.com/idaholab/moose/assets/130671073/645ca7b8-954f-4dd4-9a2c-5587ff3dfb53)
I have a boundary condition like this that I need to add, which kind of boundary should I apply? Or what kind of boundary should I inherit if I have to write it myself?

**Comment by GiudGiud** [comment]
Hello

You ll need to code a new object for that.
We do not have a ParsedDirichletBC or a MaterialDirichletBC that would also you to specify this expression

Guillaume

**Reply by jin0123456** [reply]
So I write a custom bc like this,
```
#include "ADCustomHeatFluxBC.h"

registerMooseObject("WorkerApp", ADCustomHeatFluxBC);

InputParameters
ADCustomHeatFluxBC::validParams()
{
  InputParameters params = ADIntegratedBC::validParams();
  params.addClassDescription(
      "Custom heat transfer boundary condition with temperature and heat "
      "transfer coefficent given by material properties.");
  params.addRequiredParam<MaterialPropertyName>("T_sat",
                                                "Material property for saturation temperature");
  params.addRequiredParam<MaterialPropertyName>("heat_transfer_coefficient",
                                                "Material property for heat transfer coefficient");
  return params;
}

ADCustomHeatFluxBC::ADCustomHeatFluxBC(const InputParameters & parameters)
  : ADIntegratedBC(parameters),
    _T_sat(getADMaterialProperty<Real>("T_sat")),
    _htc(getADMaterialProperty<Real>("heat_transfer_coefficient"))
{
}

ADReal
ADCustomHeatFluxBC::computeQpResidual()
{
  return -_test[_i][_qp] * _htc[_qp] * std::pow((_T_sat[_qp] - _u[_qp]), 4.);
}
```
I inherite from ADIntegratedBC, coefficient and T_sat is given by material property. However, the calculated results are different from the analytical solutions in BC, so I wonder if I inherite the wrong BC class for this expression?

**Reply by GiudGiud** [reply]
Hello

This does not look like what you pasted in your original question.
Are you implementing a heat flux or a dirichlet BC?

Guillaume

**Reply by jin0123456** [reply]
I don't know much about how to write residuals in bc but it seems not a dirichlet BC cause there is heat flux term on the right side of expression, right? See more clearly, the expression can be converted into this form,
![image](https://github.com/idaholab/moose/assets/130671073/b89bcb68-8350-4351-a11c-0b3f12729d21)
I suppose the form is like which is in `ConvectiveHeatFluxBC`, so I also inherite from ADIntegratedBC but im not sure if `computeQpResidual()` should be written the way I did?

**Reply by wangzhaohao** [reply]
Just like OP, i have the same problem about `BC`. 

> I don't know much about how to write residuals in bc but it seems not a dirichlet BC cause there is heat flux term on the right side of expression, right? See more clearly, the expression can be converted into this form,
![image](https://github.com/idaholab/moose/assets/51054734/2919eb6b-5b6a-47c5-a3e8-8ab7b497dae5)

the $q''$ is related to $T$.  So it should use `IntegralBC`? if the power of heat flux is not 1,
![image](https://github.com/idaholab/moose/assets/51054734/2aa45953-238c-4945-9d14-7593900bfbe9)

![image](https://github.com/idaholab/moose/assets/51054734/ebe49d32-2014-4599-b6bd-19bf0e88b288)

is  the convertion  suitable？

**Reply by jin0123456** [reply]
Hello, I want to ask that
1. I couldn't create a MaterialDirichletBC, right? I have tried but there is no interface for invoking material properties in a NodalBC and DirichletBC is a nodal bc, isn't it?
2. For my case, it's tend to be a IntegratedBC, right? Like the previous reply, I did something to my expression and applied in the redisual term. In this way, the calculated temperature on the boundary is lower than sateration temperature which is wrong. So I have no idea how to implement this boundary condition.
@GiudGiud Thanks very much for your reply and suggestions.

**Reply by GiudGiud** [reply]
Hello

Yes both of these are integratedBC. They model a heat flux not a fixed value of a variable. 
IntegratedBC can use material properties. 

Can you link to a github repo with your code? Maybe I can see a problem there

Irrelevant since you should not be using a DirichletBC but since you asked:
Indeed we do not support material properties on nodes so you cannot use that.
However you can:
- use functor material properties (see functors system documentation)
OR
- simply implement the expression directly in the definition of the boundary condition, which is the simplest way to do that.

**Reply by jin0123456** [reply]
https://github.com/jin0123456/bookish-garbanzo
I put files there, like I said

> I suppose the form is like which is in ConvectiveHeatFluxBC, so I also inherite from ADIntegratedBC but im not sure if computeQpResidual() should be written the way I did

I made a convertion in my expression, so the bc residual form is like ConvectiveHeatFluxBC. Like @wangzhaohao said, I also dont know if it's correct?

**Reply by GiudGiud** [reply]
Are you sure it's (T-T_sat ) ^ 4 and not T^4 - T_sat ^4 ??

The first one gives the same result if T is below or above T_sat, which seems unphysical

**Comment by jin0123456** [comment]
So it's wrong for my convertion in the expression. Before convertion, T is supposed to be above T_sat. And what's the right residual form in my case(the initial picture i pasted)?

**Reply by GiudGiud** [reply]
The initial picture you pasted is more of a dirichletBC. 
You have to choose which form you want to implement. The heat flux or the fixed temperature.

**Reply by jin0123456** [reply]
To be honest, I'm a little confused now. So what's your suggestion about which bc i should apply and how to write the residual term in my case in which the bc condition form is the initial picture i pasted? I want fixed temperature.
Thanks.

**Reply by GiudGiud** [reply]
both are correct. They are different numerical schemes that's it. 
You can implement fixed temperature with a dirichlet BC. You will want to create a new one and implement the `computeQpValue()` routine.

**Reply by jin0123456** [reply]
Oh, I see and I will try. Thanks.
